VAAPI Encoder (2nd branch/attempt)

sgothel opened the pull request:
This is the 2nd clean VAAPI branch tackling issue 1083 https://github.com/HandBrake/HandBrake/issues/1083
and superseding pull-request https://github.com/HandBrake/HandBrake/pull/3039

supported: h264, hevc, av1, untested: vp8 and vp9
- h264, hevc works on AMD >= NAVI10
- av1 works on AMD > NAVI10(?)
    
- Bitrate either VBR (variable) or CQP (constant)
    
- Using Mesa 25 / libva 2.22 (Debian 13),
  vaapi supports writing required sequence- and global-packet.
  Therefor, no more patch is required to write into a MKV container.
    
On Debian 12, Mesa 22, these packets are not written
and hence no MKV container can be used.

+++

**Tested on:**
Test environment:
    - GNU/Linux, Debian 13
    - vainfo: VA-API version: 1.22 (libva 2.22.0)
    - vainfo: Driver version: Mesa Gallium driver 25.0.7-2 for AMD Radeon RX 5700 XT (radeonsi, navi10, LLVM 19.1.7, DRM 3.61, 6.12.57+deb13-amd64)

sgothel committed:
encavcodecInit: Enable libav verbosity and show codec options if global_verbosity_level >= 3

sgothel committed:
encavcodecInit: Enable libav verbosity and show codec options if global_verbosity_level >= 3

sgothel committed:
hw-encoder-vaapi: Enables VAAPI if available and supported

supported: h264, hevc, av1, untested: vp8 and vp9
- h264 and hevc works on AMD >= NAVI10
- av1 works on AMD >= NAVI48

- Bitrate either VBR (variable) or CQP (constant)

- Using Mesa 25 / libva 2.22 (Debian 13),
  vaapi supports writing required sequence- and global-packet.
  Therefor, no more patch is required to write into a MKV container.

  On Debian 12, Mesa 22, these packets are not written
  and hence no MKV container can be used.

+++

Test environment:
- GNU/Linux, Debian 13
- vainfo: VA-API version: 1.22 (libva 2.22.0)
- vainfo: Driver version: Mesa Gallium driver 25.0.7-2 for AMD Radeon RX 5700 XT (radeonsi, navi10, LLVM 19.1.7, DRM 3.61, 6.12.57+deb13-amd64)

Codev h264_vaapi, avcodec_open options: rc_mode=CQP,qp=24,b_depth=2,profile=100,level=40 (profile high, level 4.0)
HD1080 source ~45min with 212-330fps
PSNR and SSIM difference to x264 below 1%,
having x264 using profile high, level 4.0, preset fast, rc 23.
Both using constant-quality w/ variable bitrate.

++++

If FEATURE vaapi is enable (autodetected),
the following libraries are linked: X11 va va-drm va-x11.

- hb.c:
  - Adding global static 'vaapi_device_ctx0' default device
    initialized via hb_avcodec_init() and released via hb_avcodec_free().
    The 'vaapi_device_ctx0' is used in vaa_common hb_avcodec_vaapi_set_hwframe_ctx(..).

  - hb_avcodec_vaapi_set_hwframe_ctx() attaches the new AVHWFramesContext
    to the AVCodecContext at initializing the encoder.
    This is the final sink for the vaapi hw-encoder
    and Encode(..) performs the frame conversion into it.

  - hb_avcodec_test_encoder(..) adds VAAPI branch,
    needs refinement

- encavcodec.c
  - adding VAAPI branches, enabling available, needs refinement.
    Currently profile and level are set to the user preferences.

  - added hb_avcodec_test_encoder*(..) functions, currently only
    used to check vaapi-*codec*-availability.
    We used this codefragment regularly in an earlier HB version.

  - in case of H264 w/ VAAPI, the PTS hack is disabled
    to work on AMD >= NAVI48
    - using vanilla PTS in frame and not the frame-count
    - hw_frame leaves PTS untouched

+++

Discussion:

VAAPI Workflow:
  - It uses hardware frames on the target device, which are being transported
    from the software device. The frame transport implicitly converts the pix_fmt.

  - AVCodecContext's pix_fmt uses AV_PIX_FMT_VAAPI

  - AVHWFramesContext: format uses AV_PIX_FMT_VAAPI and sw_format AV_PIX_FMT_NV12,
    the latter hinting on the actual hw-frame's target format.

  - AV_PIX_FMT_NV12 uses interleaved UV data, where AV_PIX_FMT_YUV420P uses
    seperated planes. Both use a separated Y plane upfront.
    Therefor both formats are not picture compatible, memory requirements are same.

  - Encode(..) allocates a hw_frame and the source frame is being transported
    to the hw-frame target using the pic_fmt conversion to NV12.

    Finally the hw_frame is being sent 'avcodec_send_frame' and the code-path re-aligns
    with non VAAPI.

Further fixes to do:
    add custom extra-video-encoder field

sgothel committed:
hw-encoder-vaapi: Enables VAAPI if available and supported

supported: h264, hevc, av1, untested: vp8 and vp9
- h264, hevc works on AMD >= NAVI10
- av1 works on AMD > NAVI10(?)

- Bitrate either VBR (variable) or CQP (constant)

- Using Mesa 25 / libva 2.22 (Debian 13),
  vaapi supports writing required sequence- and global-packet.
  Therefor, no more patch is required to write into a MKV container.

  On Debian 12, Mesa 22, these packets are not written
  and hence no MKV container can be used.

+++

Test environment:
- GNU/Linux, Debian 13
- vainfo: VA-API version: 1.22 (libva 2.22.0)
- vainfo: Driver version: Mesa Gallium driver 25.0.7-2 for AMD Radeon RX 5700 XT (radeonsi, navi10, LLVM 19.1.7, DRM 3.61, 6.12.57+deb13-amd64)

Codev h264_vaapi, avcodec_open options: rc_mode=CQP,qp=24,b_depth=2,profile=100,level=40 (profile high, level 4.0)
HD1080 source ~45min with 212-330fps
PSNR and SSIM difference to x264 below 1%,
having x264 using profile high, level 4.0, preset fast, rc 23.
Both using constant-quality w/ variable bitrate.

++++

If FEATURE vaapi is enable (autodetected),
the following libraries are linked: X11 va va-drm va-x11.

- hb.c:
  - Adding global static 'vaapi_device_ctx0' default device
    initialized via hb_avcodec_init() and released via hb_avcodec_free().
    The 'vaapi_device_ctx0' is used in vaa_common hb_avcodec_vaapi_set_hwframe_ctx(..).

  - hb_avcodec_vaapi_set_hwframe_ctx() attaches the new AVHWFramesContext
    to the AVCodecContext at initializing the encoder.
    This is the final sink for the vaapi hw-encoder
    and Encode(..) performs the frame conversion into it.

  - hb_avcodec_test_encoder(..) adds VAAPI branch,
    needs refinement

- encavcodec.c
  - adding VAAPI branches, enabling available, needs refinement.
    Currently profile and level are set to the user preferences,
    additionally 'b_depth=2' is being passed.

  - added hb_avcodec_test_encoder*(..) functions, currently only
    used to check vaapi-*codec*-availability.
    We used this codefragment regularly in an earlier HB version.

+++

Discussion:

VAAPI Workflow:
  - It uses hardware frames on the target device, which are being transported
    from the software device. The frame transport implicitly converts the pix_fmt.

  - AVCodecContext's pix_fmt uses AV_PIX_FMT_VAAPI

  - AVHWFramesContext: format uses AV_PIX_FMT_VAAPI and sw_format AV_PIX_FMT_NV12,
    the latter hinting on the actual hw-frame's target format.

  - AV_PIX_FMT_NV12 uses interleaved UV data, where AV_PIX_FMT_YUV420P uses
    seperated planes. Both use a separated Y plane upfront.
    Therefor both formats are not picture compatible, memory requirements are same.

  - Encode(..) allocates a hw_frame and the source frame is being transported
    to the hw-frame target using the pic_fmt conversion to NV12.

    Finally the hw_frame is being sent 'avcodec_send_frame' and the code-path re-aligns
    with non VAAPI.

Further fixes to do:
    Validate whether 'b_depth=2' (b-frames) works for all vaapi implementations,
    add custom extra-video-encoder field.

sr55 commented on libhb/handbrake/common.h:
```diff
@@ -741,6 +743,13 @@ struct hb_job_s
 
+#define HB_VCODEC_FFMPEG_VAAPI_H264  (0x00000080 | HB_VCODEC_FFMPEG_MASK | HB_VCODEC_VAAPI_MASK | HB_VCODEC_H264_MASK)
+#define HB_VCODEC_FFMPEG_VAAPI_H265  (0x00000081 | HB_VCODEC_FFMPEG_MASK | HB_VCODEC_VAAPI_MASK | HB_VCODEC_H265_MASK)
+#define HB_VCODEC_FFMPEG_VAAPI_VP8   (0x0000008a | HB_VCODEC_FFMPEG_MASK | HB_VCODEC_VAAPI_MASK)
```
Do we want to bother with VP8/9?  I'd class them as on life support these days.

Is there any AV1 support in the API?

sgothel replied:
ffmpeg -encoders | grep vaapi
```
 V....D h264_vaapi           H.264/AVC (VAAPI) (codec h264)
 V....D hevc_vaapi           H.265/HEVC (VAAPI) (codec hevc)
 V....D mjpeg_vaapi          MJPEG (VAAPI) (codec mjpeg)
 V....D mpeg2_vaapi          MPEG-2 (VAAPI) (codec mpeg2video)
 V....D vp8_vaapi            VP8 (VAAPI) (codec vp8)
 V....D vp9_vaapi            VP9 (VAAPI) (codec vp9)
```
So I guess no AV1 w/ ffmpeg version 5.1.7-0+deb12u1 at least (haven't checked 8.0.1 as used in hb)

VP8/9 .. I have no opinion here, just added the vaapi capable codecs from last patch.
Only working codec for now seems to be h264_vaapi, at least w/ AMD GPU.

sgothel replied:
I can remove the vaapi vp8/vp9, no problem. Just acknowledge.

Headcrabed replied:
seems we need 6.1+ : https://github.com/immich-app/immich/pull/10180

Headcrabed replied:
Also there are some intel gpus that support VP8/9, I advice to enable all supported vcodecs and let others to test them..

sgothel replied:
Good, guess I have to pre-test w/ my local build then. 

I propose to push the vaapi codec selection as the icing on the cake at last.

Nomis101 replied:
> ffmpeg -encoders | grep vaapi
> 
> ```
>  V....D h264_vaapi           H.264/AVC (VAAPI) (codec h264)
>  V....D hevc_vaapi           H.265/HEVC (VAAPI) (codec hevc)
>  V....D mjpeg_vaapi          MJPEG (VAAPI) (codec mjpeg)
>  V....D mpeg2_vaapi          MPEG-2 (VAAPI) (codec mpeg2video)
>  V....D vp8_vaapi            VP8 (VAAPI) (codec vp8)
>  V....D vp9_vaapi            VP9 (VAAPI) (codec vp9)
> ```
> 
> So I guess no AV1 w/ ffmpeg version 5.1.7-0+deb12u1 at least (haven't checked 8.0.1 as used in hb)
> 

AV1 was added years ago to ffmpeg 6.1
https://github.com/FFmpeg/FFmpeg/commit/3be81e3b449febfb04596f3f187aae27f18d02f7

sgothel replied:
av1_vaapi has been added (forced push)

sr55 commented on libhb/vaapi_common.c:
```diff
@@ -0,0 +1,116 @@
+    av_buffer_unref(&hw_frames_ref);
+    return err;
+#else
+    hb_log("VAAPI n/a.");
```
Not sure this log message is useful here.

sgothel replied:
Not sure either

sgothel replied:
removed

bradleysepos commented:
Thank you for this, looks pretty clean. Looking forward to again setting up Linux on my AMD machine to test (not a blocker).

galad87 commented on contrib/ffmpeg/A23-avformat-matroskaenc-tolerate-no-global-header.patch:
```diff
@@ -0,0 +1,16 @@
+     case AV_CODEC_ID_H264:
+-        return ff_isom_write_avcc(dyn_cp, extradata,
+-                                  extradata_size);
++        if (extradata_size)
```
This is a big no. It should be possible to create the extradata using a libavcodec bitstream filter

sgothel replied:
It is the same hack as for AV1 in case the header is missing (I did reference the source of the hack).
Please detail why you believe this 'is a big no', i.e. the negative impact/side-effect.
To my knowledge, this only affects the single global header and not every sequence header?

I agree though, a proper solution is desired. The referenced issue for the hack included a synthetic header for their Android meta-encoder.

For the 'proper route', please advise its implementation or/and take over this particular task.
Thank you.

galad87 replied:
H.264 extradata is required by Matroska specification. Just because some players will still be able to handle broken MKV files doesn't mean it's a good thing to do.

Check muxavformata.c:425 for something similar.

galad87 replied:
The question, is the extradata filled later, or is it never filled with the correct data?

sgothel replied:
Yes, I also don't know whether this is filled correctly using this hack attaching a blob.

Testing refactoring the extradata bitfilter (AV1) to be applied for VAAPI as well. Thank you.

sgothel replied:
I tried to analyze re-using the extradata bsf, but failed to succeed w/o the matrocaenc patch.
The latter would be even OK if `*size_to_reserve=0`, i.e. no data to be copied AFAIK.
At least it passes the `avformat_write_header`.

Whether the extradata bsf actually does anything, I am not sure - lack of knowledge here.
I could propose that commit here as an add-on, probably would need you help anyways.

sgothel replied:
I have pushed the 'hack' in a different branch, separated for discussion <https://github.com/sgothel/HandBrake/commit/4e8b16984f052e3b87e455154b75000f1346faf3>.

W/ given info that 'current' Mesa may support packet header, such a solution (if working) 
would need to be optional, best to query vaapi/ffmpeg whether packet header are supported/written?
See 
- https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3593241290
- https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3594694356

Again, I fail to properly implement this at this point due to lack of familiarity with ffmpeg's `bsf` and how it is connected in `hb`.

Alternatively, we could simply keep our eyes closed on this dirty hack - trusting it will be properly resolved with Mesa.

sgothel replied:
I was successful injecting extradata via ffmpeg by simply attaching it manually to the context
and _not_ passing the `AV_CODEC_FLAG_GLOBAL_HEADER` flag.
```
+        {
+            context->flags &= ~AV_CODEC_FLAG_GLOBAL_HEADER;
+            context->extradata_size = 1024;
+            context->extradata = (uint8_t*)av_mallocz(context->extradata_size + AV_INPUT_BUFFER_PADDING_SIZE);
+        }
```

mediainfo gave me the following delta of the resulting matroska file
```
 ID                                       : 1
 Format                                   : AVC
 Format/Info                              : Advanced Video Codec
+Format profile                           : High 10@L21.2
+MultiView_Count                          : 1 / 1
+Format settings                          : 9 Ref Frames
+Format settings, Reference frames        : 9 frames
 Codec ID                                 : V_MPEG4/ISO/AVC
 Duration                                 : 4 min 41 s
 Bit rate                                 : 5 206 kb/s
 Width                                    : 1 920 pixels
+Original width                           : 12 pixels
 Height                                   : 800 pixels
-Display aspect ratio                     : 2.40:1
+Original height                          : 32 pixels
+Display aspect ratio                     : 0.375
 Frame rate mode                          : Constant
 Frame rate                               : 23.976 (24000/1001) FPS
 Color space                              : YUV
+Chroma subsampling                       : 4:2:0
+Bit depth                                : 8 bits
+Scan type                                : Progressive
 Bits/(Pixel*Frame)                       : 0.141
 Stream size                              : 175 MiB (85%)
 Default                                  : Yes
```

while ffprobe showed the additional extradata size
```
-extradata_size=36
+extradata_size=1024
```

I will prepare this patch and resolve this last issue - before adding av1_vaapi to finalize it.

galad87 replied:
That's random data. As you can see mediainfo is showing crazy things.

sgothel replied:
hmm, OK. Thx, now that you mention it :). 
Was hopeful since ffmpeg's matroskaenc received the extradata_size and the code-path writing was active.

The bfs filter hack didn't inject any data either, so **I still need your help** here.
Otherwise, we could leave it as-is w/ the ffmpeg's matroskaenc patch just claiming zero additional-space.
This leads to the extradata_size of 36 (probably padding)

sgothel replied:
OK only revised(*) the ffmpeg matroskaenc patch, reserving zero bytes and resulting in tolerating missing extradata and giving an extradata_size of 36 bytes in the mkv output (ffprobe).
Mediainfo stays short.

Edit: Added same for hvec/h265 so all three will work w/ av1_vaapi and an older Mesa library.
Using the new Mesa, shall result in proper header/extradata.
I will leave this as-is now.

(*) forced push

sgothel replied:
I vote for marking this issue resolved. Perhaps mentioning this MKV issue w/ vaapi for older Mesa driver, resulting in potential incompatible MKV files (which I couldn't verify here w/ more lax open-source tools).

sgothel replied:
I have removed the questionable ffmpeg patch, as MKV files are working w/ Debian 13 / Mesa 25
see https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3639952958

galad87 commented on libhb/encavcodecaudio.c:
```diff
@@ -251,6 +251,22 @@ static int encavcodecaInit(hb_work_object_t *w, hb_job_t *job)
     // packet instead.
     context->flags |= AV_CODEC_FLAG_GLOBAL_HEADER;
 
+    if( 3 <= global_verbosity_level ) {
```
Seems unrelated, might want to move it to a different PR, and fix the code style.

galad87 commented on libhb/encavcodec.c:
```diff
@@ -1264,14 +1468,44 @@ static void Encode( hb_work_object_t *w, hb_buffer_t **buf_in,
         frame.flags = 0;
     }
 
-    // Encode
```
Is this needed? All the other libavcodec hardware encoders don't require a manual hw upload.

sgothel replied:
I annotated this pix_fmt conversion for vaapi in the commit message (discussion). 
If it can be avoided in current HB, i.e. using same pix_fmt throughout filters etc - I would love to drop this extra work.

```
VAAPI Workflow:
      - It uses hardware frames on the target device, which are being transported
        from the software device. The frame transport implicitly converts the pix_fmt.
    
      - AVCodecContext's pix_fmt uses AV_PIX_FMT_VAAPI
    
      - AVHWFramesContext: format uses AV_PIX_FMT_VAAPI and sw_format AV_PIX_FMT_NV12,
        the latter hinting on the actual hw-frame's target format.
    
      - AV_PIX_FMT_NV12 uses interleaved UV data, where AV_PIX_FMT_YUV420P uses
        seperated planes. Both use a separated Y plane upfront.
        Therefor both formats are not picture compatible, memory requirements are same.
    
      - Encode(..) allocates a hw_frame and the source frame is being transported
        to the hw-frame target using the pic_fmt conversion to NV12.
    
        Finally the hw_frame is being sent 'avcodec_send_frame' and the code-path re-aligns
        with non VAAPI.
```

galad87 commented on libhb/encavcodec.c:
```diff
@@ -1812,3 +2062,102 @@ const int* hb_av_get_pix_fmts(int encoder)
      }
 }
+
+static int avcodec_test_encoder(const AVCodec *codec, enum AVPixelFormat fmt)
```
Should be in hbffmpeg.c

galad87 commented on libhb/handbrake/common.h:
```diff
@@ -741,6 +743,13 @@ struct hb_job_s
+#define HB_VCODEC_FFMPEG_VAAPI_VP8   (0x0000008a | HB_VCODEC_FFMPEG_MASK | HB_VCODEC_VAAPI_MASK)
+#define HB_VCODEC_FFMPEG_VAAPI_VP9   (0x0000008b | HB_VCODEC_FFMPEG_MASK | HB_VCODEC_VAAPI_MASK)
+
+#define HB_VCODEC_IS_VAAPI(a)        ( 0 != ( (a) & HB_VCODEC_VAAPI_MASK ) )
```
If these masks are unused, it would be better to just not defined them.

sgothel replied:
they are used!

galad87 commented on libhb/hb.c:
```diff
@@ -28,6 +28,11 @@
 
+#if HB_PROJECT_FEATURE_VAAPI
+static const char * vaapi_device0 = NULL; // "/dev/dri/renderD128";
+AVBufferRef *vaapi_device_ctx0 = NULL; // reused in vaapi_common.c
```
Please don't add any global variable. It should be possible to run multiple encodes in a single process.

sgothel replied:
`vaapi_device_ctx0` is used for `av_hwframe_ctx_alloc` and it should be thread-safe, if initialized statically as-is.

However, we could hide it in the vaapi space to sooth style concerns.
Not familiar with the (new) encoder modularization, but if it provides a static init/free callback, sure.

galad87 commented on libhb/hb.c:
```diff
@@ -76,6 +81,30 @@ int disable_hardware = 0;
 
 static void thread_func( void * );
 
+static void hb_avcodec_init()
```
Check how the other hardware encoders are implemented. This shouldn't be needed, and if it's needed, it shouldn't be a global variable or be here.

sgothel replied:
The whole avcodec_init() came from the earlier versions where same style was used, impact is only as used - static initialization.
(Same w/ above remarks)

Maybe I can find time to change this to the current hacking style.

sgothel commented:
I have force pushed this branch, addressed some comments.

rathann commented:
> This is the 2nd clean VAAPI branch tackling issue 1083 #1083 and superseding pull-request #3039

Thanks for keeping at it!

> h264 works on AMD >= NAVI10, unsupported/untested: h265, vp8 and vp9

I've got a NAVI22, which supports h264 and h265 encoding. I'll try to test this soon.

>     * Bitrate either VBR (variable) or CQP (constant)
> 
> 
> This approach works w/o patched Mesa gallium va frontend, instead it uses a ffmpeg/avformat/matrokaenc patch allowing h264 streams w/o sequence nor global header. See https://trac.ffmpeg.org/ticket/10667.

FWIW, the current [code in mesa](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/frontends/va/config.c?ref_type=heads#L301) looks like this:
```C
         case VAConfigAttribEncPackedHeaders:
            value = VA_ENC_PACKED_HEADER_NONE;
            if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_MPEG4_AVC))
               value |= ENC_PACKED_HEADERS_H264;
            else if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_HEVC))
               value |= ENC_PACKED_HEADERS_HEVC;
            else if (u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_AV1)
               value |= ENC_PACKED_HEADERS_AV1;
```
and the `ENC_PACKED_HEADERS_*` constants referenced above do include `VA_ENC_PACKED_HEADER_SEQUENCE | VA_ENC_PACKED_HEADER_SLICE | VA_ENC_PACKED_HEADER_MISC`.

To me, this means mesa doesn't need to be patched anymore, right?

sgothel commented:
> Thanks for keeping at it!
.. and thank you for keep looking at it too.

> I've got a NAVI22, which supports h264 and h265 encoding. I'll try to test this soon.
Great. Thx.

> FWIW, the current [code in mesa](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/frontends/va/config.c?ref_type=heads#L301) looks like this:
> 
> ```c
>          case VAConfigAttribEncPackedHeaders:
>             value = VA_ENC_PACKED_HEADER_NONE;
>             if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_MPEG4_AVC))
>                value |= ENC_PACKED_HEADERS_H264;
>             else if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_HEVC))
>                value |= ENC_PACKED_HEADERS_HEVC;
>             else if (u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_AV1)
>                value |= ENC_PACKED_HEADERS_AV1;
> ```
> 
> and the `ENC_PACKED_HEADERS_*` constants referenced above do include `VA_ENC_PACKED_HEADER_SEQUENCE | VA_ENC_PACKED_HEADER_SLICE | VA_ENC_PACKED_HEADER_MISC`.
> 
> To me, this means mesa doesn't need to be patched anymore, right?

Yes, looks like!

Mesa commit <https://gitlab.freedesktop.org/mesa/mesa/-/commit/fc4abbe27d063337a420d147cf8c9fa492789f71>,
not sure which release that is (from July 2024).

So we may need to unblind vaapi in ffmpeg public API, allowing to query these values?
At least w/ my (old) Debian 12.12 version of my test machine this change didn't sooth ffmpeg yet,
still claiming no packet headers produced. Will have a check here.

sgothel commented:
Force pushed w/ added `av1_vaapi` support (untested due to lack of testing hardware, like `hevc_vaapi`)

Mastergatto commented:
Hi, I own a Radeon RX 9070 XT so I'm able to test this PR with H.264, HEVC and AV1. 

First of all, here is what I get with `vainfo` on Linux:
```
Trying display: wayland
vainfo: VA-API version: 1.22 (libva 2.22.0)
vainfo: Driver version: Mesa Gallium driver 25.2.7-arch1.1 for AMD Radeon RX 9070 XT (radeonsi, gfx1201, LLVM 21.1.5, DRM 3.64, 6.17.9-arch1-1)
vainfo: Supported profile and entrypoints
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileHEVCMain10             :	VAEntrypointEncSlice
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointVLD
      VAProfileAV1Profile0            :	VAEntrypointVLD
      VAProfileAV1Profile0            :	VAEntrypointEncSlice
      VAProfileNone                   :	VAEntrypointVideoProc
```
As you can see, VAAPI works on my system and my GPU can encode with H.264, HEVC and AV1, but not with VP9.

The command line used for this test is: `HandBrakeCLI -i 'video.ogg' -o '<codec>.mkv' -e <encoder parameter>` where `video.ogg` is a sample video with a length of 8 seconds, is codified with Ogg/Theora and has a framerate of 25 fps. File size is 2,9 MB. Here's what I have found by trying to transcode it to H.264, HEVC and AV1.

H.264:
Despite it saying the encode has failed, it has converted the video although the result is so-so: the new video is effectively a slideshow going at 8 fps, and its size is reduced to 773 kB which doesn't make much sense.
https://pastebin.com/3KV5W2aY

HEVC:
The conversion is successful, the result is really good. The video size is reduced to 1,3 MB.
https://pastebin.com/BrgeStK1

AV1:
The conversion has been a failure. And to nitpick regarding the parameter name for flag -e (for Handbrake's CLI), for transcoding with AV1 through VAAPI should be 'vaapi_av1' not 'vaapi_avi1'.
https://pastebin.com/1y3St0gV

As a control test, I have also tried to transcode with ffmpeg, using VAAPI, and it has worked flawlessly for each codec.

sgothel commented:
Thank you!

> for transcoding with AV1 through VAAPI should be 'vaapi_av1' not 'vaapi_avi1'
fixed typo

> Hi, I own a Radeon RX 9070 XT so I'm able to test this PR with H.264, HEVC and AV1.
Great! I just figured I also have at least HEVC encoding available, just not AV1.

> 
> First of all, here is what I get with `vainfo` on Linux:
> 
> ```
> Trying display: wayland
> vainfo: VA-API version: 1.22 (libva 2.22.0)
> vainfo: Driver version: Mesa Gallium driver 25.2.7-arch1.1 for AMD Radeon RX 9070 XT (radeonsi, gfx1201, LLVM 21.1.5, DRM 3.64, 6.17.9-arch1-1)
> vainfo: Supported profile and entrypoints
> ```
>
> As you can see, VAAPI works on my system and my GPU can encode with H.264, HEVC and AV1, but not with VP9.

Yup!
 
> The command line used for this test is: `HandBrakeCLI -i 'video.ogg' -o '<codec>.mkv' -e <encoder parameter>` where `video.ogg` is a sample video with a length of 8 seconds, is codified with Ogg/Theora and has a framerate of 25 fps. File size is 2,9 MB. Here's what I have found by trying to transcode it to H.264, HEVC and AV1.
> 
> H.264: Despite it saying the encode has failed, it has converted the video although the result is so-so: the new video is effectively a slideshow going at 8 fps, and its size is reduced to 773 kB which doesn't make much sense. https://pastebin.com/3KV5W2aY

I could reproduce this here also with ffmpeg transcoding, stuttering/jumping frames from theora,
but not from like AV1 or H264.

My test script:
```
LOGL="-loglevel 40"

ffmpeg -y ${LOGL} -ss 05:00 -vaapi_device /dev/dri/renderD128 -i m0.mkv -t 00:20 -vf format=nv12,hwupload -c:v h264_vaapi -c:a copy m0_h264_vaapi.mp4 2>&1 | tee m0_h264_vaapi.log
ffmpeg -y ${LOGL} -ss 05:00 -vaapi_device /dev/dri/renderD128 -i m0.mkv -t 00:20 -vf format=nv12,hwupload -c:v hevc_vaapi -c:a copy m0_hevc_vaapi.mp4 2>&1 | tee m0_hevc_vaapi.log
ffmpeg -y ${LOGL} -ss 05:00 -vaapi_device /dev/dri/renderD128 -i m0.mkv -t 00:20 -vf format=nv12,hwupload -c:v av1_vaapi -c:a copy m0_av1_vaapi.mp4 2>&1 | tee m0_av1_vaapi.log
ffmpeg -y ${LOGL} -ss 05:00 -i m0.mkv -t 00:20 -c:v libtheora -c:a opus -strict -2 m0_theora.ogg 2>&1 | tee m0_theora.log
ffmpeg -y ${LOGL} -ss 05:00 -i m0.mkv -t 00:20 -an -c:v libtheora m0_theora.aonly.ogg 2>&1 | tee m0_theora.aonly.log

IN_AV=m0_theora.ogg
IN_V=m0_theora.aonly.ogg

ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_AV} -vf format=nv12,hwupload -c:v h264_vaapi -c:a copy m0_theora_h264_vaapi.av.mp4 2>&1 | tee m0_theora_h264_vaapi.av.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_V} -vf format=nv12,hwupload -c:v h264_vaapi m0_theora_h264_vaapi.v.mp4 2>&1 | tee m0_theora_h264_vaapi.v.log
```

The last two `m0_theora_h264_vaapi.av.mp4` and `m0_theora_h264_vaapi.v.mp4` show the bug.
Same w/ hevc/h265 vaapi from theora.

Notable, I use the same ffmpeg 8.0.1 version (manual build) as our current Handbrake build
`commit 894da5ca7d742e4429ffb2af534fcda0103ef593 (HEAD -> b_8_0_1, tag: n8.0.1)`

```
ffmpeg version n8.0.1 Copyright (c) 2000-2025 the FFmpeg developers
  built with gcc 12 (Debian 12.2.0-14+deb12u1)
  configuration: --enable-gpl --enable-shared --enable-runtime-cpudetect --prefix=/opt-linux-x86_64/ffmpeg-8.0.1 --enable-libmp3lame --enable-libtwolame --enable-libx264 --enable-libx265 --enable-libsvtav1 --enable-libvpx --enable-libtheora --enable-encoder=h264_nvenc --enable-encoder=hevc_nvenc --enable-encoder=h264_vaapi --enable-encoder=hevc_vaapi --enable-encoder=av1_vaapi --enable-encoder=vp8_vaapi --enable-encoder=vp9_vaapi
  libavutil      60.  8.100 / 60.  8.100
  libavcodec     62. 11.100 / 62. 11.100
  libavformat    62.  3.100 / 62.  3.100
  libavdevice    62.  1.100 / 62.  1.100
  libavfilter    11.  4.100 / 11.  4.100
  libswscale      9.  1.100 /  9.  1.100
  libswresample   6.  1.100 /  6.  1.100
```

This is potentially an ffmpeg issue ...

> HEVC: The conversion is successful, the result is really good. The video size is reduced to 1,3 MB. https://pastebin.com/BrgeStK1
> 
> AV1: The conversion has been a failure. And to nitpick regarding the parameter name for flag -e (for Handbrake's CLI), for transcoding with AV1 through VAAPI should be 'vaapi_av1' not 'vaapi_avi1'. https://pastebin.com/1y3St0gV
> 
> As a control test, I have also tried to transcode with ffmpeg, using VAAPI, and it has worked flawlessly for each codec.

The HEVC + AV1 detection is also fixed, I was using  h264 only parameter in `hb_avcodec_test_encoder`.
Thank you for pointing this out.

Will force push soon after some additional tests.

Mastergatto commented:
With the proposed changes above, encode with av1 through VAAPI now works and with good results, though it still complains about missing profile:
```
[av1_vaapi @ 0x5621d0bfed80] No usable encoding profile found.
[20:40:58] hb_avcodec_test_encoder_available encoder=0x40018082: err -3
```
And a warning:
`[20:40:58] encavcodecInit: Unknown avcodec option q`

Also, while with CLI every codec works, but in the GUI only `H.264 (vaapi) `is available.

Mastergatto commented:
> This is potentially an ffmpeg issue ...

Not sure about the bug being in ffmpeg. I use the same version, 8.0.1.

In my case transcoding directly with ffmpeg was flawless with each codec, it was done with a command line like this, e.g. for H.264:  `ffmpeg -vaapi_device /dev/dri/renderD128 -i video.ogg -vf 'format=nv12,hwupload' -c:v h264_vaapi h264.mkv`

sgothel commented:
@Mastergatto OK pushed revision w/ our noted changes

Below my test scripts.

Tested w/ below scripts and a proper theora test file (my bad),
no stuttering etc. 

Couldn't test AV1, due to hardware - but h264 + hevc/h265 works OK w/ vaapi here.

If you test w/ HB, please ensure the `--verbose=4` commandline,
which will enable ffmpeg verbose logging of the vaapi core.

+++

cat ffmpeg_test.sh 
```
#!/bin/sh

LOGL="-loglevel 40"
IN=m1_h264.mp4

ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v h264_vaapi -c:a copy m1_h264_vaapi.mp4 2>&1 | tee m1_h264_vaapi.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v hevc_vaapi -c:a copy m1_hevc_vaapi.mp4 2>&1 | tee m1_hevc_vaapi.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v av1_vaapi -c:a copy m1_av1_vaapi.mp4 2>&1 | tee m1_av1_vaapi.log

ffmpeg -y -i ${IN} -c:v libtheora -r 25 -qscale:v 10 -c:a opus -strict -2 m1_theora.ogg 2>&1 | tee m1_theora.log
ffmpeg -y -i ${IN} -an -c:v libtheora -r 25 -qscale:v 10 -strict -2 m1_theora.v.ogg 2>&1 | tee m1_theora.v.log

IN_TAV=m1_theora.ogg
IN_TV=m1_theora.v.ogg

ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_TAV} -vf format=nv12,hwupload -c:v h264_vaapi -c:a copy m1_theora_h264_vaapi.av.mp4 2>&1 | tee m1_theora_h264_vaapi.av.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_TV} -vf format=nv12,hwupload -c:v h264_vaapi m1_theora_h264_vaapi.v.mp4 2>&1 | tee m1_theora_h264_vaapi.v.log

ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_TAV} -vf format=nv12,hwupload -c:v hevc_vaapi -c:a copy m1_theora_h265_vaapi.av.mp4 2>&1 | tee m1_theora_h265_vaapi.av.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN_TV} -vf format=nv12,hwupload -c:v hevc_vaapi m1_theora_h265_vaapi.v.mp4 2>&1 | tee m1_theora_h265_vaapi.v.log
```

cat hb_test.sh 
```
#!/bin/sh

IN=m1_h264.mp4
IN_TAV=m1_theora.ogg
IN_TV=m1_theora.v.ogg

HB=/opt-linux-x86_64/HandBreak/bin/HandBrakeCLI
HB_OPTS="--verbose=4 --no-dvdnav"

${HB} ${HB_OPTS} -i ${IN} -o 'hb_x264_to_h264.mkv' -e vaapi_h264 2>&1 | tee hb_x264_to_h264.log
${HB} ${HB_OPTS} -i ${IN} -o 'hb_x264_to_h265.mkv' -e vaapi_hevc 2>&1 | tee hb_x264_to_h265.log

${HB} ${HB_OPTS} -i ${IN_TAV} -o 'hb_theora_to_h264.av.mkv' -e vaapi_h264 2>&1 | tee hb_theora_to_h264.av.log
${HB} ${HB_OPTS} -i ${IN_TAV} -o 'hb_theora_to_h265.av.mkv' -e vaapi_hevc 2>&1 | tee hb_theora_to_h265.av.log

${HB} ${HB_OPTS} -i ${IN_TV} -o 'hb_theora_to_h264.v.mkv' -e vaapi_h264 2>&1 | tee hb_theora_to_h264.v.log
${HB} ${HB_OPTS} -i ${IN_TV} -o 'hb_theora_to_h265.v.mkv' -e vaapi_hevc 2>&1 | tee hb_theora_to_h265.v.log
```

sgothel commented:
> With the proposed changes above, encode with av1 through VAAPI now works and with good results, though it still complains about missing profile:
> 
> ```
> [av1_vaapi @ 0x5621d0bfed80] No usable encoding profile found.
> [20:40:58] hb_avcodec_test_encoder_available encoder=0x40018082: err -3
> ```
> 
> And a warning: `[20:40:58] encavcodecInit: Unknown avcodec option q`
> 
> Also, while with CLI every codec works, but in the GUI only `H.264 (vaapi) `is available.

I revised the `hb_avcodec_test_encoder` and added `av1_vaapi` encoder in the ffmpeg module w/ last forced push,
should be visible in UI and working.

Mastergatto commented:
H.264, still not good:
https://pastebin.com/M9PbtSH5
I have tried to convert other videos (Theora, AV1 and HEVC) to H.264, still the same result. I have also tried with a much longer video, Handbrake only converted the first 30 seconds and skipped the rest.

For what's worth, I  have also tested AV1 - OK:
https://pastebin.com/npZ1gvjS

And HEVC - OK:
https://pastebin.com/akN9UGYY

For GUI, now all three VAAPI encoders are available in the menu, and they produce same result as the CLI.

Thank you for your hard work!

sgothel commented:
> available



> H.264, still not good: https://pastebin.com/M9PbtSH5 I have tried to convert other videos (Theora, AV1 and HEVC) to H.264, still the same result. I have also tried with a much longer video, Handbrake only converted the first 30 seconds and skipped the rest.

Hmm, weird .. as it works for me :|

What I see is this in your log
```
[matroska @ 0x7f6445e1ed00] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 120 >= 40
ERROR: avformatMux: track 0, av_interleaved_write_frame failed with error 'Invalid argument'
[matroska @ 0x7f6445e1ed00] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 120 >= 80
```

So this may need some more elaboration, could be due to 
our different Mesa/vaapi implementation or hardware.

+++

Notable, your new Mesa/vaapi produces the packed header (our dire debate above), good.
(see last line - shows the packet header extradata size by vaapi)

```
[23:35:11] encavcodec: avcodec_open options: rc_mode=CQP,qp=22,b_depth=2
[h264_vaapi @ 0x7f6445cf82c0] Input surface format is nv12.
[h264_vaapi @ 0x7f6445cf82c0] Using VAAPI profile VAProfileH264High (7).
[h264_vaapi @ 0x7f6445cf82c0] Using VAAPI entrypoint VAEntrypointEncSlice (6).
[h264_vaapi @ 0x7f6445cf82c0] Using VAAPI render target format YUV420 (0x1).
[h264_vaapi @ 0x7f6445cf82c0] RC mode: CQP.
[h264_vaapi @ 0x7f6445cf82c0] Block Level bitrate control: OFF.
[h264_vaapi @ 0x7f6445cf82c0] RC quality: 22.
[h264_vaapi @ 0x7f6445cf82c0] RC framerate: 25/1 (25.00 fps).
[h264_vaapi @ 0x7f6445cf82c0] Driver does not report any additional prediction constraints.
[h264_vaapi @ 0x7f6445cf82c0] Using intra, P- and B-frames (supported references: 1 / 1).
[h264_vaapi @ 0x7f6445cf82c0] All wanted packed headers available (wanted 0xd, found 0x1f).
[23:35:11] Writing Metadata to output file...
[matroska @ 0x7f6445e1ed00] mkv_assemble: h264: extradata_size 46
```

> 
> For what's worth, I have also tested AV1 - OK: https://pastebin.com/npZ1gvjS
> 
> And HEVC - OK: https://pastebin.com/akN9UGYY
> 
> For GUI, now all three VAAPI encoders are available in the menu, and they produce same result as the CLI.

Great.

> 
> Thank you for your hard work!

Thank you!

sgothel commented:
force pushed changing inline function `hb_vcodec_is_vaapi` to regular function `hb_video_encoder_is_vaapi`, as required to build w/ debugging options `--debug=std --optimize=none`.

sgothel commented:
Forced pushed w/o sub-optimal ffmpeg patch:
- Using Mesa 25 / libva 2.22 (Debian 13),
  vaapi supports writing required sequence- and global-packet.
  Therefor, no more patch is required to write into a MKV container.
    
  On Debian 12, Mesa 22, these packets are not written
  and hence no MKV container can be used.
    
_Edit_:
- test machine _vainfo: Driver version: Mesa Gallium driver 25.0.7-2 for AMD Radeon RX 5700 XT (radeonsi, navi10, LLVM 19.1.7, DRM 3.61, 6.12.57+deb13-amd64)_
- on same machine w/ same test file: fps ~177 -> ~350 w/ vaapi h264 or h265
- h264 conversion is not corrupted (see earlier discussion w/ @Mastergatto https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3621339145)

galad87 commented on libhb/hb.c:
```diff
@@ -2163,6 +2164,8 @@ int hb_global_init()
         return -1;
     }
 
+    hb_vaapi_init();
```
Should be done when checking for availability like it's done for the other hardware encoders.

sgothel replied:
See https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3676655837

galad87 commented on libhb/vaapi_common.c:
```diff
@@ -0,0 +1,111 @@
+#if HB_PROJECT_FEATURE_VAAPI
+    {
+        int err;
+        if( (err = av_hwdevice_ctx_create(&vaapi_device_ctx0, AV_HWDEVICE_TYPE_VAAPI,
```
What if there are multiple devices?

sgothel replied:
Good point, thx. 
I briefly looked at the _device index_ (or similar, to selecting a device from multiple available),
but didn't yet follow through this abstraction - namely the assignment and compilation of available devices.
Default is _-99_ AFAIK .. but I can look at this enhancement later on. 
If you - or anybody else - has some pointers/docs for this multi-device HB feature, please share.

galad87 commented on libhb/encavcodec.c:
```diff
@@ -1264,14 +1518,48 @@ static void Encode( hb_work_object_t *w, hb_buffer_t **buf_in,
-    // Encode
-    ret = avcodec_send_frame(pv->context, &frame);
+#if HB_PROJECT_FEATURE_VAAPI
+    if (NULL != pv->context->hw_frames_ctx &&
```
It would be better to move the hw_upload to a separate hw_upload filter.

sgothel replied:
I like to encourage you to be a little bit more verbose.
- a separate ffmpeg filter chain stage? Please give me an HB example.
- or just a C-function for better encapsulation?

galad87 replied:
A wrapper over an AVFilter, like the format filter in format.c. Or just move it to a function for now. encavcodec.c is such a mess.

sgothel replied:
I looked at AVFilter to add the `nv12,hwupload`, but failed to 
- find a proper AVFILTER_FLAG_HWDEVICE flagged AVFilter
- where to attach out vaapi hw-context 

However, I isolated the vaapi `hwupload` in `hb_vaapi_avcodec_send_frame`
and also made sure `av_frame_unref(&frame)` get issued.

galad87 reviewed (commented):
Please use the same coding style as the rest of libhb:

blah *foo;
if (foo)
{
    printf("hello");
}

sgothel commented:
Changes:
- See https://github.com/HandBrake/HandBrake/pull/7467#discussion_r2636320759
- Also moved 
  - `hb_vaapi_init` into `hb_common_global_hw_init`
  - `hb_vaapi_free` into new `hb_common_global_hw_close`
 
Selection of vaapi device has not been addressed here yet,
perhaps a detail for a followup patch?
This could involve a potential UI change to select and exceeds my time right now.
Edit: Alternatively, we could use an environment variable to optionally select a device path.

HexStan commented:
Hi, thanks for your hard work! I tried converting a few videos to HEVC, but they all glitched out. My GPU is the Vega 8 integrated into the R7-5800H. It features VCN 2.2, which is higher than the VCN 2.0 found in Navi 10, so I believe it should fully meet the hardware requirements.

https://pastebin.com/bqE7jZ7m

https://pastebin.com/4tmAj46U

<img width="1280" height="720" alt="720p x265-vaapi qp29 opus mp4_20251221_163129 080" src="https://github.com/user-attachments/assets/797dacba-56a7-4249-a6b9-d2998f2a4862" />

AVC works fine.

sgothel commented:
> Hi, thanks for your hard work! I tried converting a few videos to HEVC, but they all glitched out. My GPU is the Vega 8 integrated into the R7-5800H. It features VCN 2.2, which is higher than the VCN 2.0 found in Navi 10, so I believe it should fully meet the hardware requirements.
> 
> https://pastebin.com/bqE7jZ7m
> 
> https://pastebin.com/4tmAj46U

Please
- run w/ environment variable `VERBOSE=99` or `-loglevel 4` (or higher), showing us ffmpeg verbose + debug messages
- show your `vainfo` output.
- also see https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3621266816 (*) where one user also showed similar issue with a higher version Mesa driver and AMD GPU.
- I suspect some issues w/ ffmpeg, mesa and/or our encoding configuration
  - try uncommenting `libhb/encavcodec.c` line 1023 `av_dict_set(&av_opts, "b_depth", "2", 0);` and report
> 
> AVC works fine.

Setup 1 (mine)
```
- GNU/Linux, Debian 13
- vainfo: VA-API version: 1.22 (libva 2.22.0)
- vainfo: Driver version: Mesa Gallium driver 25.0.7-2 for AMD Radeon RX 5700 XT (radeonsi, navi10, LLVM 19.1.7, DRM 3.61, 6.12.57+deb13-amd64)

- H264: OK
- HEVC: OK
- AV1: not available
```
Setup 2(*)
```
VA-API version: 1.22 (libva 2.22.0)
vainfo: Driver version: Mesa Gallium driver 25.2.7-arch1.1 for AMD Radeon RX 9070 XT (radeonsi, gfx1201, LLVM 21.1.5, DRM 3.64, 6.17.9-arch1-1)

- H264: buggy
- HEVC: OK
- AV1: OK
```

HexStan commented:
> Please
> 
> * run w/ environment variable `VERBOSE=99` or `-loglevel 4` (or higher), showing us ffmpeg verbose + debug messages
> * show your `vainfo` output.
> * also see [VAAPI Encoder (2nd branch/attempt) #7467 (comment)](https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3621266816) (*) where one user also showed similar issue with a higher version Mesa driver and AMD GPU.
> * I suspect some issues w/ ffmpeg, mesa and/or our encoding configuration
>   
>   * try uncommenting `libhb/encavcodec.c` line 1023 `av_dict_set(&av_opts, "b_depth", "2", 0);` and report

Here is my `vainfo` output:

```
libva info: VA-API version 1.20.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_20
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.20 (libva 2.12.0)
vainfo: Driver version: Mesa Gallium driver 25.0.7-0ubuntu0.24.04.2 for AMD Radeon Graphics (radeonsi, renoir, ACO, DRM 3.57, 6.8.0-90-generic)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileVC1Simple              : VAEntrypointVLD
      VAProfileVC1Main                : VAEntrypointVLD
      VAProfileVC1Advanced            : VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointEncSlice
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointEncSlice
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointEncSlice
      VAProfileHEVCMain               : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointEncSlice
      VAProfileHEVCMain10             : VAEntrypointVLD
      VAProfileHEVCMain10             : VAEntrypointEncSlice
      VAProfileJPEGBaseline           : VAEntrypointVLD
      VAProfileVP9Profile0            : VAEntrypointVLD
      VAProfileVP9Profile2            : VAEntrypointVLD
      VAProfileNone                   : VAEntrypointVideoProc
```

Regarding `libhb/encavcodec.c`, I checked the source code and noticed that the line `av_dict_set(&av_opts, "b_depth", "2", 0);` was actually **not commented out** in the original code.

Here are the logs running with the **original code** (where `b_depth` is active):
[https://pastebin.com/aHVm2XFQ](https://pastebin.com/aHVm2XFQ)

I tried **commenting out** `av_dict_set(&av_opts, "b_depth", "2", 0);` to test, but the output video is still glitchy. Visually, the glitches appear exactly the same as before, and the bitrate is nearly identical. Here are the logs for this test:
[https://pastebin.com/90FF3yRV](https://pastebin.com/90FF3yRV)

On the other hand, encoding to AVC and HEVC directly using the `ffmpeg` CLI works perfectly fine. Here are the commands and logs for that:
[https://pastebin.com/WMvfviA3](https://pastebin.com/WMvfviA3)

sgothel commented:
> Regarding `libhb/encavcodec.c`, I checked the source code and noticed that the line `av_dict_set(&av_opts, "b_depth", "2", 0);` was actually **not commented out** in the original code.

My bad, yes - *comment out*, my cold makes me a bit fuzzy :)
(So no change here, good to know - thx)

Thank you for testing.

+++

Tricky  .. You and the other tester use a more up-to-date GPU than me.
Your Ubuntu version is almost on par w/ mine, while the other one is more up-to-date w/ arch.

As you both wrote, the build in vanilla `ffmpeg 6.1.1-3ubuntu5` produces flawless conversion, 
our HB setup w/ 8.0.1 (edited) does not. 
- coding error on my behalf only impacting newer cards (sync?)
  - the Encode block is pretty straight forward though?
- ffmpeg version + GPU version + GPU firmware issue?

HexStan commented:
> My bad, yes - _comment out_, my cold makes me a bit fuzzy :) (So no change here, good to know - thx)
> 
> Thank you for testing.
> 
> +++
> 
> Tricky .. You and the other tester use a more up-to-date GPU than me. Your Ubuntu version is almost on par w/ mine, while the other one is more up-to-date w/ arch.
> 
> As you both wrote, the build in vanilla `ffmpeg 6.1.1-3ubuntu5` produces flawless conversion, our HB setup w/ 8.0.1 (edited) does not.
> 
> * coding error on my behalf only impacting newer cards (sync?)
>   
>   * the Encode block is pretty straight forward though?
> * ffmpeg version + GPU version + GPU firmware issue?

Get well soon! I actually just recovered from a cold myself, so I know the feeling.

I just tried using a static build of FFmpeg 8.0.1 to see if I could reproduce the issue outside of HandBrake. However, I encountered a symbol resolution error:
`implib-gen: libva.so.2: failed to resolve symbol 'vaMapBuffer2' via dlsym: /lib/x86_64-linux-gnu/libva.so.2: undefined symbol: vaMapBuffer2`

Instead of the visual glitches I saw earlier, this resulted in a 0kb output mp4 file.

This might be unrelated to your current investigation (likely just a library version mismatch on my system), but I'm sharing the logs just in case they are useful to you:

https://pastebin.com/ZCVx0imf

HexStan commented:
> Tricky .. You and the other tester use a more up-to-date GPU than me. Your Ubuntu version is almost on par w/ mine, while the other one is more up-to-date w/ arch.
> 
> As you both wrote, the build in vanilla `ffmpeg 6.1.1-3ubuntu5` produces flawless conversion, our HB setup w/ 8.0.1 (edited) does not.
> 
> * coding error on my behalf only impacting newer cards (sync?)
>   
>   * the Encode block is pretty straight forward though?
> * ffmpeg version + GPU version + GPU firmware issue?

Uh oh, new finding: outputting to `.mkv` seems to work fine.

sgothel commented:
Below is a summary of our above tests, including @HexStan `mkv` note https://github.com/HandBrake/HandBrake/pull/7467#issuecomment-3689605452
[HBIssues-va_api.md](https://github.com/user-attachments/files/24341027/HBIssues-va_api.md)

Perhaps we can merge this PR as-is and create a followup 
to address the remaining issues w/ below test machines 
regarding `h264` and `mp4` container?

### Test Machine Setup

- Machine 1 (author)
  - Mesa Gallium driver 25.0.7-2
    - LLVM 19.1.7
    - DRM 3.61
  - Navi10
    - AMD Radeon RX 5700 XT
  - Linux kernel 6.12.57+deb13-amd64

- Machine 2 (mastergatto)
  - Mesa Gallium driver 25.2.7-arch1.1
    - LLVM 21.1.5
    - DRM 3.64
  - gfx1201
    - AMD Radeon RX 9070 XT
  - Linux kernel 6.17.9-arch1-1
  - Works w/ ffmpeg 8.0.1 cli

- Machine 3 (HexStan)
  - Mesa Gallium driver 25.0.7-0ubuntu0.24.04.2
    - ACO
    - DRM 3.57
  - Renoir
    - AMD Radeon Graphics R7-5800H (Vega 8)
  - Linux kernel 6.8.0-90-generic)
  - Works w/ ffmpeg 6.0.1 cli

### Test Results

- X: corrupted encoding output
- ?: untested


| Machine | h264 mp4 | h264 mkv | h265 mp4 | h265 mkv | av1 mp4 | av1 mkv |
| ------: | :------: | :------: | :------: | :------: | :-----: | :-----: |
|       1 |    ok    |    ok    |    ok    |    ok    |   n/a   |   n/a   |
|       2 |    ?     |    X     |    ?     |    ok    |   ?     |   ok    |
|       3 |    ?     |    ?     |    X     |    ok    |   X     |   ok    |

rathann commented:
FWIW, h264 encode fails for me with all samples I tried. I get the following errors:
```
[15:17:14] avformatMux: track 0, av_interleaved_write_frame failed with error 'Invalid argument'
```
and then:
```
[matroska @ 0x7ff490058180] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 100 >= 67
[matroska @ 0x7ff490058180] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 100 >= 83
[matroska @ 0x7ff490058180] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 150 >= 117
...
```
Same errors with mp4 output, too.

I'm on Fedora 43 with Radeon RX 6700 (navi22), Mesa 25.2.7, FFmpeg 7.1.2, VA-API 1.22/libva 2.22.0, LLVM 21.1.5, DRM 3.64 and kernel 6.17.12.

h264_vaapi encode works via ffmpeg command line just fine.

h265 encode works, though CPU usage is 500-600% on an 8-thread CPU. It is ~10 times faster than x265 at "fast" preset with similar CPU usage on my machine.

av1 encode is not supported by this GPU.

nfinite-recursion commented:
| GPU | h264 MP4 | h264 MKV | h265 MP4 | h265 MKV | AV1 MP4 | AV1 MKV |
|--------|--------|--------|--------|--------|--------|--------|
| 680M | Failed / Garbled / 32s | Failed / Garbled / 32s | Garbled | OK | N/A | N/A |
| 7700S | Failed / Garbled / 32s | Failed / Garbled / 32s | Garbled | OK | OK | OK |
| 780M | Failed / 32s | Failed / 32s | Garbled | OK | OK | OK |

Hardware/Software Notes:

680M (GFX1035/Rembrandt) / 7700S (GFX1102/Navi33)

- Mesa Gallium driver 25.1.5-1pop0~1756399231~22.04~b84bab8
- LLVM 15.0.7
- DRM 3.64
- 6.17.4-76061704-generic
- PopOS 22.04

780M (GFX1103/Phoenix)

- Mesa Gallium driver 25.3.0
- LLVM 21.1.6
- DRM 3.64
- 6.17.7-ba20.fc43.x86_64
- Bazzite 43

---

- This git branch was built in a Distrobox instance using Arch Linux on both systems.
- Both configurations used FFMPEG 8.0.1.
- VAAPI hardware encoding directly with FFMPEG passed on all hardware configurations no matter the codec.
- H.264 VAAPI hardware encoding with Handbrake failed on all hardware configurations, encoding only the first 32 seconds of my test video before stopping. However, the 680M/7700S configuration had garbled output whereas the 780M configuration was clear but the playback framerate was slow and choppy, despite VLC and other players reporting correct framerate. I'll note that the 780M configuration uses a newer version of Mesa and indeed, there was a colorspace fix in the Mesa version immediately after version 25.1.5 found on the 680M/7700S configuration.
- All failed H.264 encoding logs feature a "ERROR: avformatMux: track 0, av_interleaved_write_frame failed with error 'Invalid argument'" line, suggesting the output is not handled properly by Handbrake. Not sure if this is an upstream issue or if the Handbrake main branch needs to be modified.
- H.265 in an MP4 container doesn't fail, but is always garbled; MKV is always fine.
- AV1 always passes no matter which container so long as the hardware supports it.

I also modified and combined the example scripts above:

cat vaapi_test.sh
```
#!/bin/sh

LOGL="-loglevel 40"
IN=input.mp4
OUT_DIR=./output/

mkdir ${OUT_DIR}

ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v h264_vaapi -c:a copy ${OUT_DIR}ffmpeg_h264_vaapi.mp4 2>&1 | tee ${OUT_DIR}ffmpeg_h264_vaapi.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v hevc_vaapi -c:a copy ${OUT_DIR}ffmpeg_hevc_vaapi.mp4 2>&1 | tee ${OUT_DIR}ffmpeg_hevc_vaapi.log
ffmpeg -y ${LOGL} -vaapi_device /dev/dri/renderD128 -i ${IN} -vf format=nv12,hwupload -c:v av1_vaapi -c:a copy ${OUT_DIR}ffmpeg_av1_vaapi.mp4 2>&1 | tee ${OUT_DIR}ffmpeg_av1_vaapi.log

HB=~/Downloads/HandBrake/build/HandBrakeCLI
HB_OPTS="--verbose=4 --no-dvdnav"

${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_h264.mkv" -e vaapi_h264 2>&1 | tee ${OUT_DIR}hb_x264_to_h264_mkv.log
${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_h265.mkv" -e vaapi_hevc 2>&1 | tee ${OUT_DIR}hb_x264_to_h265_mkv.log
${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_av1.mkv" -e vaapi_av1 2>&1 | tee ${OUT_DIR}hb_x264_to_av1_mkv.log

${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_h264.mp4" -e vaapi_h264 2>&1 | tee ${OUT_DIR}hb_x264_to_h264_mp4.log
${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_h265.mp4" -e vaapi_hevc 2>&1 | tee ${OUT_DIR}hb_x264_to_h265_mp4.log
${HB} ${HB_OPTS} -i ${IN} -o "${OUT_DIR}hb_x264_to_av1.mp4" -e vaapi_av1 2>&1 | tee ${OUT_DIR}hb_x264_to_av1_mp4.log
```

sgothel commented:
Thank you @nfinite-recursion for your detailed report!

Most (?) conversions seem to work with native `ffmpeg`, indeed indicating issues with the 
avmuxer code in HB (Did I miss something? Some vaapi incompatibility?).

Then there also might be issues with the native vaapi H264 encoding on the newer AMD cards,
since my old one works OK.

Question to the maintainer: How shall we proceed?
If willing to merge this feature, I would need to finalize the merge conflicts first.
If not willing to merge, we must leave this sitting here for another while :)

sr55 commented:
I'd rather not let this rot so it's probably better we take this in and it can be off by default until we narrow it down.

@galad87 @bradleysepos any objections?

galad87 commented:
Code style is still wrong in many places, and the check needs to be the same as the other hardware encoder, the hw_device context has to be created only when needed, and then released.

sgothel commented:
(1) Code Style

> Code style is still wrong in many places, 

In my last update, I corrected a few code style edits manually.
Do we have a clang-format config file, which would allow us to apply it pre-checkin?
Of course, code-style often differs a little - also in the HB code base.

2) hw-device context lock

> and the check needs to be the same as the other hardware encoder, the hw_device context has to be created only when needed, and then released.

Hmm, I could go ahead an pick the NV hw-encoder and follow its code-path?
If this is sufficient, so be it - otherwise please give me a better recommendation.

3) hw-device selection

Open question here is the selection of the hw-device itself as commented above.
This can be already done via an environment variable .. or needs to be determined more properly
in a subsequent commit (allowing to select suitable devices).

+++

Regardless of the last item, I will look into item (2) at least.
For item (1) I need a clang-format file to fix potentially remaining style diversions (ahem).

galad87 commented:
It surely differs, but new code should at least be consistent, something like:

blabla *foo = whatever;
if (foo)
{
     do_thing();
}

sgothel commented:
So .. [GNU formatting](https://www.gnu.org/prep/standards/standards.html#Formatting) I assume?
We can use `GNU` for [clang-format BasedOnStyle](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#basedonstyle)'s. OK.

sr55 committed:
Merge branch 'master' into hw-encoder-vaapi2
